sklearn/multioutput.py

```diff
@@ -359,3 +362,229 @@ def score(self, X, y):
+            Returns self.
+        """
+
+        random_state = check_random_state(self.random_state)
+
+        if self.order is None:
+            self.order = list(range(Y.shape[1]))
+        elif self.order == 'random':
+            self.order = random_state.permutation(Y.shape[1])
+        elif sorted(self.order) != list(range(Y.shape[1])):
+                raise ValueError("invalid order")
+
+        self.estimators_ = [clone(self.base_estimator)
+                            for _ in range(Y.shape[1])]
+
+        self.classes_ = []
+
+        if sp.issparse(X):
+            if self.cv is None:
+                for chain_idx, estimator in enumerate(self.estimators_):
+                    if chain_idx == 0:
+                        X_aug = X
+                    else:
+                        X_aug = sp.hstack((X, Y[:, self.order[:chain_idx]]))
+                    y = Y[:, self.order[chain_idx]]
+                    estimator.fit(X_aug, y)
+                    self.classes_.append(estimator.classes_)
+            else:
+                Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
+                for chain_idx, estimator in enumerate(self.estimators_):
+                    if chain_idx == 0:
+                        X_aug = X
+                    else:
+                        X_aug = sp.hstack((X, Y[:, self.order[:chain_idx]]))
+                    y = Y[:, self.order[chain_idx]]
+                    estimator.fit(X_aug, y)
+                    Y_pred_chain[:, chain_idx] = cross_val_predict(
+                        self.base_estimator, X, y=y, cv=self.cv)
+                    self.classes_.append(estimator.classes_)
+
+        else:
+            if self.cv is None:
+                X_aug = np.hstack((X, Y[:, self.order]))
+                for chain_idx, estimator in enumerate(self.estimators_):
+                    y = Y[:, self.order[chain_idx]]
+                    estimator.fit(X_aug[:, :(X.shape[1] + chain_idx)], y)
+                    self.classes_.append(estimator.classes_)
+
+            else:
+                Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
+                for chain_idx, estimator in enumerate(self.estimators_):
+                    X_aug = np.hstack((X, Y_pred_chain[:, :chain_idx]))
+                    y = Y[:, self.order[chain_idx]]
+                    estimator.fit(X_aug, y)
+                    Y_pred_chain[:, chain_idx] = cross_val_predict(
+                        self.base_estimator, X, y=y, cv=self.cv)
+                    self.classes_.append(estimator.classes_)
+
+    def predict(self, X):
+        """Predict on the data matrix X using the ClassifierChain model.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+
+        Returns
+        -------
+        Y_pred : array-like, shape (n_samples, n_classes)
+        """
+
+        Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
```

Comment:
You can again hstack X to this and fill it in as you go to avoid repeated copying.